### ISSUE TITLE
Some issues about the diffusion model.

### ISSUE THREAD

**Issue by SiriYu-2001** [Melmaphother/TimeDART#6]
I have read the paper roughly, and tried to debug the program for some times. I wonder that if there any diffusion sampling exists? I didn't find any codes such as DDPM, DDIM or any other diffusion sampling processes. So what's the affect of diffusion? If it's just a replacement of MSELoss? By the way, I find there are no diffusion forward processes when you are fine-tuning and test the result. Could you answer my question? Thank you very much.

    def forecast(self, x):
        batch_size, _, num_features = x.size()
        if self.use_norm:
            means = torch.mean(x, dim=1, keepdim=True).detach()
            x = x - means
            stdevs = torch.sqrt(
            torch.var(x, dim=1, keepdim=True, unbiased=False) + 1e-5
            ).detach()
            x = x / stdevs

        x = self.channel_independence(x)  # [batch_size * num_features, input_len, 1]
        x = self.patch(x)  # [batch_size * num_features, seq_len, patch_len]
        x = self.enc_embedding(x)  # [batch_size * num_features, seq_len, d_model]
        x = self.positional_encoding(x)  # [batch_size * num_features, seq_len, d_model]
        x = self.encoder(
            x,
            is_mask=False,
        )  # [batch_size * num_features, seq_len, d_model]
        x = x.reshape(
            batch_size, num_features, -1, self.d_model
        )  # [batch_size, num_features, seq_len, d_model]

        # forecast
        x = self.head(x)  # [bs, pred_len, n_vars]

        # denormalization
        if self.use_norm:
            x = x * (stdevs[:, 0, :].unsqueeze(1)).repeat(1, self.pred_len, 1)
            x = x + (means[:, 0, :].unsqueeze(1)).repeat(1, self.pred_len, 1)

        return x

**Comment by Melmaphother** [comment]
> I have read the paper roughly, and tried to debug the program for some times. I wonder that if there any diffusion sampling exists? I didn't find any codes such as DDPM, DDIM or any other diffusion sampling processes. So what's the affect of diffusion? If it's just a replacement of MSELoss? By the way, I find there are no diffusion forward processes when you are fine-tuning and test the result. Could you answer my question? Thank you very much.我粗略地阅读了这篇论文，并尝试了几次调试该程序。我想知道是否存在任何扩散采样？我没有找到任何代码，例如 DDPM、DDIM 或任何其他扩散采样过程。那么扩散有什么影响呢？如果它只是 MSELoss 的替代品？顺便说一句，我发现在微调和测试结果时没有扩散前向过程。您能回答我的问题吗？谢谢。
> 
> ```
> def forecast(self, x):
>     batch_size, _, num_features = x.size()
>     if self.use_norm:
>         means = torch.mean(x, dim=1, keepdim=True).detach()
>         x = x - means
>         stdevs = torch.sqrt(
>         torch.var(x, dim=1, keepdim=True, unbiased=False) + 1e-5
>         ).detach()
>         x = x / stdevs
> 
>     x = self.channel_independence(x)  # [batch_size * num_features, input_len, 1]
>     x = self.patch(x)  # [batch_size * num_features, seq_len, patch_len]
>     x = self.enc_embedding(x)  # [batch_size * num_features, seq_len, d_model]
>     x = self.positional_encoding(x)  # [batch_size * num_features, seq_len, d_model]
>     x = self.encoder(
>         x,
>         is_mask=False,
>     )  # [batch_size * num_features, seq_len, d_model]
>     x = x.reshape(
>         batch_size, num_features, -1, self.d_model
>     )  # [batch_size, num_features, seq_len, d_model]
> 
>     # forecast
>     x = self.head(x)  # [bs, pred_len, n_vars]
> 
>     # denormalization
>     if self.use_norm:
>         x = x * (stdevs[:, 0, :].unsqueeze(1)).repeat(1, self.pred_len, 1)
>         x = x + (means[:, 0, :].unsqueeze(1)).repeat(1, self.pred_len, 1)
> 
>     return x
> ```

Thank you for your attention to our work. This question is also a question asked by the previous reviewer, which reminds us that even if we have revised the article, it still confuses readers (probably because the paradigm of diffusion sampling such as DDPMs has been deeply rooted in people's minds and maybe my pool writing 😂). This is our shortcoming. Next, please let me explain it carefully:
First of all, TimeDART is a **self-supervised** framework for pre-training to fine-tuning **rather than** a supervised framework. You can take a look at Figure 1 below, which clearly shows the different processes of pre-training and fine-tuning.
In pre-training, we trained both the causal Transformer encoder and the masked denoising decoder, but downstream we only migrated the Transformer encoder, removed the causal mask, and did not migrate the denoising decoder. It is a two-stage training process.
Let me explain your doubts below:
1. is there any diffusion sampling exists?
No, self-supervised task is to better train the **encoder** network in pre-training through a carefully designed pre-training paradigm, and then migrate the encoder to the downstream, connect different heads to complete different downstream tasks (for example in computer vision, the training process is like [MAE](https://arxiv.org/abs/2111.06377)); instead of doing time series generation tasks (forecasting and generation are easily confused, you can see the [Diffusion-TS](https://arxiv.org/abs/2403.01742) article for this point)
2. Why don't we keep diffusion sampling?
Because the sampling process is too **slow** and our downstream tasks are not generation but forecasting and classification. The effect of using sampling for forecasting tasks will be very poor and slow, that is, the **MSE will be extremely high** (for experimental proof of this conclusion, you can see another article [FDF](https://arxiv.org/abs/2410.13253)). It is a bit different from standard DDPMs that could makes your confusion.
3. what's the affect of diffusion? If it's just a replacement of MSELoss?
From the above explanation, you could be able to find that we only use the diffusion-denoising process to train the encoder and denoising decoder in pre-training but not fine-tuning? Why do we only introduce the diffusion process in pre-training? This involves the biggest innovation of our article in the pre-training paradigm: we do not mechanically combine the autoregressive and diffusion models, but use the diffusion denoising process to explicitly introduce noise in pre-training, enhance the robustness of autoregressive optimization, and perform fine-grained modeling at the patch-level, rather than a simple MLP or Conv1D mapping. The goal of pre-training is diffusion loss (the 8th formula in the article, it **looks very similar ** to MSE loss, because diffusion loss was originally in the form of cross entropy, and finally formed the form of MSE. The detailed derivation here can be seen at [derivation at my own github page](https://melmaphother.github.io/2024/08/08/Denoising-Diffusion-Probabilistic-Model/#more)).
As for downstream tasks, forecasting uses MSE loss and classification uses cross entropy. This is a unified operation of the self-supervised method. In order to ensure fairness, we are consistent with other frameworks (for pre-training and fine-tuning, you can see the [TimeSiam](https://arxiv.org/abs/2402.02475)).
4. I find there are no diffusion forward processes when you are fine-tuning and test the result?
Of course, there is no, because we have removed the denoising decoder in the downstream task and no longer need forward diffusion. The downstream task is **also consistent with other self-supervised methods**.

In a word, the introduction of denoising decoder, diffusion, and diffsion loss are all to optimize the encoder network in pre-training, so as to learn more useful and robust representations, so that they can be better optimized in downstream tasks, that is, better than random init.

Once again, we apologize for the misunderstanding caused by our unclear article. If you have any other questions, please feel free to discuss.

Figure 1

![Image](https://github.com/user-attachments/assets/6f6c6691-7748-4207-921d-eaa572bceb12)
